docs: keptn Scheduler architecture documentation (#1777)
keptn/lifecycle-toolkit: docs/content/en/docs/concepts/architecture/components/scheduler/_index.md

cascade:
---

> The **Keptn Scheduler**, an integral component of the Keptn Lifecycle Toolkit which orchestrates

Reviewer: ```suggestion
The **Keptn Scheduler** is an integral component of the Keptn Lifecycle Toolkit that orchestrates
```

the deployment process.

> The **Keptn Scheduler** works by registering itself as a Permit plugin with Kubernetes that

Reviewer 2: ```suggestion
The **Keptn Scheduler** works by registering itself as a Permit plugin within the Kubernetes scheduling cycle
```

> is responsible for ensuring that Pods are scheduled to a node until and unless the

Reviewer: ```suggestion
that ensures that Pods are scheduled to a node until and unless the
```
"...is responsible for..." is just extra noise -- I like the tighter prose.

pre-deployment checks have finished successfully.

> This is important because it helps to prevent Pods from being scheduled to nodes that are

Reviewer 2: ```suggestion
This helps to prevent Pods from being scheduled to nodes that are
```

> not yet ready for them, which can lead to errors and downtime

Reviewer 2: ```suggestion
not yet ready for them, which can lead to errors and downtime. Furthermore, it allows the user to control the deployment of an app based on customized rules that can take into consideration more parameters than what the default scheduler has (simple CPU and memory parameters)
```

## How does the Keptn Scheduler works

> The Keptn Scheduler first checks the annotations on the Pod to see if it is annotated with

Reviewer 2: The mutating webhook checks the annotations, and assigns the keptn scheduler only if they are present. So the Keptn scheduler only gets the pods that are annotated.

[Keptn specific annotations](https://main.lifecycle.keptn.sh/docs/implementing/integrate/#basic-annotations).
If the Pod is annotated with Keptn specific annotations, the Keptn Scheduler retrieves
the WorkloadInstance CRD that is associated with the Pod.
The **WorkloadInstance CRD** contains information about the `pre-deployment` checks that
need to be performed before the Pod can be scheduled.

The Keptn Scheduler then checks the status of the WorkloadInstance CRD to see if the
`pre-deployment` checks have finished successfully.
If the pre-deployment checks have finished successfully, the **Keptn Scheduler** allows

> the Pod to be scheduled to a node.

Reviewer 2: I would also add here that since the scheduler is a plugin to the default k8s scheduler, all other checks that are usually performed by it apply here, for instance, if no node has enough capacity for the deployment this will not be scheduled whether our pre-deployment phase succeded or not.

If the `pre-deployment` checks have not yet finished, the Keptn Scheduler tells Kubernetes to check again later.

The Keptn Scheduler processes the following information from the WorkloadInstance CRD:

- The name of the pre-deployment checks that need to be performed.
- The status of the pre-deployment checks.

> - The deadline for the pre-deployment checks to be completed.

Reviewer 2: shall we add here the time? I believe the scheduler checks every 10 sec and has a 5 minutes upperlimit

Reviewer: So how does one modify the time configuration?

The Keptn Scheduler uses this information to determine whether or not the Pod can be scheduled.
If all of the pre-deployment checks have finished successfully and the deadline has not been reached,
the Keptn Scheduler allows the Pod to be scheduled.
If any of the `pre-deployment` checks have not finished successfully or the deadline has
been reached, the Keptn Scheduler tells Kubernetes to check again later.

> The Keptn Scheduler also has the ability to blacklist nodes that have failed pre-deployment checks.

Reviewer 2: I do not agree with this sentence, this may be a k8s default behavior. (not sure) but we do not black list nodes. I would remove this part

This means that the Keptn Scheduler will not schedule Pods to nodes that have failed pre-deployment
checks in the past.
This helps to prevent Pods from being scheduled to nodes that are not ready for them.

> The **Keptn Scheduler** is a powerful tool that can be used to ensure that Pods are
> only scheduled to nodes that are ready for them.
> This helps to prevent errors and downtime, and it also helps to improve the overall reliability of your Kubernetes cluster.

Reviewer 2: ```suggestion

```
Since this does not give new information I would prefer to remove it

> This project uses the

Reviewer 2: I would probably move this information to the top, when you mention that the scheduler is a plugin 😸

cascade:
---

The **Keptn Scheduler** is an integral component of the Keptn Lifecycle Toolkit that orchestrates
the deployment process.
The **Keptn Scheduler** works by registering itself as a Permit plugin within the Kubernetes
scheduling cycle that ensures that Pods are scheduled to a node until and unless the
pre-deployment checks have finished successfully.
This helps to prevent Pods from being scheduled to nodes that are not yet ready for them,
which can lead to errors and downtime.
Furthermore, it also allows users to control the deployment of an application based on
customized rules that can take into consideration more parameters than what the default
scheduler has (typically CPU and memory values).

The Keptn Scheduler uses the Kubernetes
[Scheduler Framework](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/) and is based on the

> [Scheduler Plugins Repository](https://github.com/kubernetes-sigs/scheduler-plugins/tree/master)

Reviewer 3: nit
```suggestion
[Scheduler Plugins Repository](https://github.com/kubernetes-sigs/scheduler-plugins/tree/master).
```

> 

Reviewer 3: maybe worth mentioning that we are using a permit plugin 
https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#permit

Author: Sure will add that info

deployment, testing, and remediation at specific times or intervals.
The Keptn Scheduler guarantees that Pods are initiated only after
the Pre-Deployment checks are completed.

## Architecture of Keptn Scheduler

To integrate the `Keptn Lifecycle Scheduler` into a Kubernetes cluster,
a Kubernetes manifest is applied with specific **Keptn annotations**.
These annotations identify the workload and specify the
pre-deployment and post-deployment tasks and evaluations

> associated with it.

Reviewer 2: This is correct more or less, but I would rephrase: if a workload/deployment is annotated with Keptn annotations the webhook will edit it to refer to the Keptn scheduler instead of the default k8s one

## How it works

The `Keptn Scheduler`, integrated into the `Keptn Lifecycle Operator`,
which is equipped with multiple controllers for managing

> **Keptn Custom Resource Definitions (CRDs)** and a powerful

Reviewer: ```suggestion
**Keptn resources** and a powerful
```
To clarify the terminology, a CRD (Custom Resource Definition) yaml file/manifest describes the structure of our resources (KeptnApp/KeptnWorkload, etc...), while a CR manifest describes a instances of those resources.  We also have Kubernetes resources like Deployment, Namespace, etc.  We are gravitating to just calling these "resources" in text for the most part.

> The `Keptn Scheduler`, integrated into the `Keptn Lifecycle Operator`,
> which is equipped with multiple controllers for managing
> **Keptn Custom Resource Definitions (CRDs)** and a powerful
> **Mutating Webhook**.
> Its core responsibility is to orchestrate the scheduling of Pods, taking
> into account pre-deployment checks to ensure optimal deployment success.
> 
> Let's dive into its architectural details:
> 

Reviewer 2: ```suggestion

```

- **Scheduler Flow**:
After applying a Kubernetes Manifest annotated with Keptn-specific

> annotations to the cluster, the Mutating Webhook injects the Keptn Scheduler.

Reviewer: This should refer to the "Basic Keptn annotations or labels" and be a link to https://main.lifecycle.keptn.sh/docs/implementing/integrate/#basic-annotations , which describes these annotations.

The Scheduler then receives Kubernetes Events for

> Pre-Deployment, which are sent to the event stream.

Reviewer: Be careful here...  With just the basic annotations, you get the observability features but do not get the pre- and post-deployment evaluations and tasks -- those are separate.  The existing documentation is a bit misleading here -- PR coming soon that attempts to correct it.  Apparently the annotations/labels for the evaluations and tasks are applied to the Kubernetes Workloads and affect only the KeptnWorkloads individually.  The pre- and post-deployment evaluations and tasks are added manually to the KeptnApp resource by name,   People may run the observability features without using the release lifecycle management features (which are the tasks/evaluations applied to the entire KeptnApp so we need to describe here how the Scheduler manages each different piece.

The Event Controller watches for these events and triggers a
Kubernetes Job to fulfill the Pre-Deployment tasks.
Once the Pre-Deployment phase is complete,

> the Keptn Scheduler schedules the Pod for deployment.

Reviewer 2: This is not exactly how it works, the scheduler is triggered every time a new workload is added, it will wait for the workload instance pre-evaluation status to be successful before binding each pod to a node. Please refer to this https://github.com/keptn/lifecycle-toolkit/blob/main/README.md#scheduler

- **Controllers**:
The Keptn Lifecycle Operator contains several controllers
for `Keptn Custom Resource Definitions (CRDs)` and a `Mutating Webhook`.
These controllers, including **KeptnApp** and **KeptnWorkload**,
monitor and manage workload resources and checks associated with an application.

    ```yaml
    apiVersion: v1
    kind: Namespace
    metadata:
      name: podtato-kubectl
      annotations:
        keptn.sh/lifecycle-toolkit: "enabled"  # this lines tells the webhook to handle the    namespace
        keptn.sh/workload: "some-workload-name"

>         keptn.sh/app: "your-app-name"

Reviewer: Verify with someone else but I'm pretty sure that these annotations are applied to one of the workload resources and not to the Namespace.  Only the `enabled` annotation is added to the Namespace resource to enable KLT in that namespace.

    ```

- **KeptnApp**:
The [KeptnApp](https://lifecycle.keptn.sh/docs/yaml-crd-ref/app/) represents information about all workloads
and checks associated with an application.
It specifies the pre/post deployment and evaluation checks
to be executed at the app level.
Changing the app version triggers new executions of
app-level checks, while changes in the workload version
impact only workload checks.

    ```yaml
    apiVersion: lifecycle.keptn.sh/v1alpha3
    kind: KeptnApp
    metadata:
      name: podtato-head
      namespace: podtato-kubectl
    spec:
      version: "1.3"
      revision: 1
      workloads:
        - name: podtato-head-left-arm
          version: 0.1.0
        - name: podtato-head-left-leg
          version: 1.2.3
      preDeploymentTasks:
        - pre-deployment-hello
      postDeploymentTasks:
        - post-deployment-hello
      preDeploymentEvaluations:
        - my-prometheus-definition
      postDeploymentEvaluations:

>         - my-prometheus-definition

Reviewer: I'm not crazy about reproducing the whole resource definition in the Architecture section.  The reference page for the KeptnApp resource has this and it is showing up elsewhere.  The Architecture docs are not the same thing as the how-to and reference docs.

    ```

> - **KeptnWorkload**:

Reviewer: For this section, each resource should be a link to the reference page in the https://main.lifecycle.keptn.sh/docs/yaml-crd-ref/ section or, if it is not there, in the https://main.lifecycle.keptn.sh/docs/crd-ref/ section.

is a CRD used to define tasks that the Keptn Lifecycle Toolkit can run
during pre- and post-deployment phases.
Task definitions can be created in the **application's namespace** or the **default** KLT namespace.

- **KeptnTask**:
A Task is responsible for executing the `TaskDefinition` of a workload.
It spawns a Kubernetes Job to handle a single Task and keeps track of the Job's status.

- **KeptnEvaluationDefinition**:
The [KeptnEvaluationDefinition](https://lifecycle.keptn.sh/docs/yaml-crd-ref/evaluationdefinition/)
is a CRD used to define evaluation tasks that the Keptn Lifecycle Toolkit
runs during pre- and post-analysis phases.
Like `KeptnTaskDefinition`, it can be created in the application's namespace or the default KLT namespace.

  A KeptnEvaluationDefinition looks like the following:

  ```yaml
   apiVersion: lifecycle.keptn.sh/v1alpha3
    kind: KeptnEvaluationDefinition
    metadata:
      name: my-prometheus-evaluation
    spec:
      objectives:
        - keptnMetricRef:
            name: cpu
            namespace: ns
          evaluationTarget: <20
        - keptnMetricRef:
            name: memory
            namespace: ns
          evaluationTarget: >4
  ```

> 

Reviewer: Again, this is on the reference page and does not belong in the Architecture section

> It keeps track of currently active `Workload Instances` responsible
> for performing these checks for specific instances of Deployments,
> StatefulSets, or ReplicaSets.
> 
> - **KeptnWorkloadInstance**:
> A Workload Instance **executes** `Pre-Deployment` and `Post-Deployment`
> checks for a specific workload.
> It maintains the status of all checks and the overall state of the `Pre-Deployment` phase.
> Once the desired state is reached (e.g., all pods are up and running), a `PostDeploymentCheck` can be
> triggered.
> 
>   For specifying the Pre-Deployment and Post-Deployment checks to be
>   executed for the Workload, use the following annotations:
> 
>    `keptn.sh/pre-deployment-tasks: task1,task2`
> 
>    `keptn.sh/post-deployment-tasks: task1,task2`
> 
>   Similarly, for defining the Evaluations, use the annotations:
> 
>    `keptn.sh/pre-deployment-evaluations: my-evaluation-definition`
> 
>    `keptn.sh/post-deployment-evaluations: my-eval-definition`
> 
> - **KeptnTaskDefinition**:
> The [KeptnTaskDefinition](https://lifecycle.keptn.sh/docs/yaml-crd-ref/taskdefinition/)
> is a CRD used to define tasks that the Keptn Lifecycle Toolkit can run
> during pre- and post-deployment phases.
> Task definitions can be created in the **application's namespace** or the **default** KLT namespace.
> 
> - **KeptnTask**:
> A Task is responsible for executing the `TaskDefinition` of a workload.
> It spawns a Kubernetes Job to handle a single Task and keeps track of the Job's status.
> 
> - **KeptnEvaluationDefinition**:
> The [KeptnEvaluationDefinition](https://lifecycle.keptn.sh/docs/yaml-crd-ref/evaluationdefinition/)
> is a CRD used to define evaluation tasks that the Keptn Lifecycle Toolkit
> runs during pre- and post-analysis phases.
> Like `KeptnTaskDefinition`, it can be created in the application's namespace or the default KLT namespace.
> 
>   A KeptnEvaluationDefinition looks like the following:
> 
>   ```yaml
>    apiVersion: lifecycle.keptn.sh/v1alpha3
>     kind: KeptnEvaluationDefinition
>     metadata:
>       name: my-prometheus-evaluation
>     spec:
>       objectives:
>         - keptnMetricRef:
>             name: cpu
>             namespace: ns
>           evaluationTarget: <20
>         - keptnMetricRef:
>             name: memory
>             namespace: ns
>           evaluationTarget: >4
>   ```
> 
> ## Conculsion

Reviewer 2: ```suggestion

```

> The Keptn Scheduler, together with other components of the Keptn
> Lifecycle Toolkit, revolutionizes Kubernetes scheduling and deployment
> by enabling event-driven **tasks** and **actions** throughout the application lifecycle.
> It empowers users with greater control, automation, and observability,
> ultimately ensuring smoother and more reliable application deployments.

Reviewer 2: ```suggestion

```
